linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala

```diff
@@ -0,0 +1,403 @@
+       * work out of the box). In the absence of any other specific place, an
+       * Element section with the declarative mode is the recommended way to
+       * introduce these declarations.
+       */
+      val exprs = funcDeclarations.map { funcID =>
+        wa.Expr(List(wa.RefFunc(funcID)))
+      }
+      ctx.moduleBuilder.addElement(
+        wamod.Element(watpe.RefType.funcref, exprs, wamod.Element.Mode.Declarative)
+      )
+    }
+  }
+
+  private def buildJSFileContent(module: ModuleSet.Module,
+      wasmFileName: String, importedModules: List[String]): String = {
+    val (moduleImports, importedModulesItems) = (for {
+      (moduleName, idx) <- importedModules.zipWithIndex
+    } yield {
+      val identName = s"imported$idx"
+      val escapedModuleName = "\"" + moduleName + "\""
+      val moduleImport = s"import * as $identName from $escapedModuleName"
+      val item = s"  $escapedModuleName: $identName,"
+      (moduleImport, item)
+    }).unzip
+
+    val (exportDecls, exportSetters) = (for {
+      exportName <- module.topLevelExports.map(_.exportName)
+    } yield {
+      val identName = s"exported$exportName"
+      val decl = s"let $identName;\nexport { $identName as $exportName };"
+      val setter = s"  $exportName: (x) => $identName = x,"
+      (decl, setter)
+    }).unzip
+
+    s"""
+      |${moduleImports.mkString("\n")}
+      |
+      |import { load as __load } from './${config.loaderModuleName}';
+      |
+      |${exportDecls.mkString("\n")}
+      |
+      |await __load('./${wasmFileName}', {
+      |${importedModulesItems.mkString("\n")}
+      |}, {
+      |${exportSetters.mkString("\n")}
+      |});
+    """.stripMargin.trim() + "\n"
+  }
+}
+
+object Emitter {
+
+  /** Configuration for the Emitter. */
+  final class Config private (
+      val coreSpec: CoreSpec,
+      val loaderModuleName: String
+  )
+
+  object Config {
+    def apply(coreSpec: CoreSpec, loaderModuleName: String): Config =
+      new Config(coreSpec, loaderModuleName)
+  }
+
+  final class Result(
+      val wasmModule: wamod.Module,
+      val loaderContent: Array[Byte],
+      val jsFileContent: String
+  )
+
+  /** Builds the symbol requirements of our back-end.
+   *
+   *  The symbol requirements tell the LinkerFrontend that we need these symbols to always be
+   *  reachable, even if no "user-land" IR requires them. They are roots for the reachability
+   *  analysis, together with module initializers and top-level exports. If we don't do this, the
+   *  linker frontend will dead-code eliminate our box classes.
+   */
+  private def symbolRequirements(coreSpec: CoreSpec): SymbolRequirement = {
+    val factory = SymbolRequirement.factory("wasm")
+
+    factory.multiple(
+      factory.instantiateClass(ClassClass, ClassCtor),
+
+      // TODO Ideally we should not require this, but rather adapt to its absence
+      factory.instantiateClass(JSExceptionClass, JSExceptionCtor),
```

Comment:
I didn't expect `JSExceptionClass` to be in symbol requirements, why is this required? (what problem gonna be there without this?)

Reply:
It turned necessary for the `helloworld` build to work. I think it only happens if the codebase contains a `try..xatch` but no `throw`.

Reply:
> try..catch but no throw.

ah, that makes sense, thanks!